src/com/google/cloud/datastore/types.js

```diff
@@ -0,0 +1,312 @@
+          booleanValue: foam.Boolean,
+          integerValue: foam.Number,
+          doubleValue: foam.Number,
+          timestampValue: foam.Date,
+          stringValue: foam.String,
+          entityValue: foam.core.FObject
+        };
+        return function typeOfDatastoreValue(v) {
+          for ( var key in v ) {
+            if ( typeMap[key] && v.hasOwnProperty(key) ) return typeMap[key];
+          }
+
+          throw new Error('Failed to identify type of datastore value: ' +
+              JSON.stringify(v));
+        };
+      })()
+  ]
+});
+
+//
+// Refine foam.core.FObject to support datastore's:
+// (1) "kinds" (i.e., types);
+// (2) "keys" (i.e., <type, name|auto-generated-id> pairs);
+// (3) "entities" (i.e., objects);
+// (4) "values" (i.e., primitives or entities).
+//
+
+foam.CLASS({
+  refines: 'foam.core.FObject',
+
+  methods: [
+    function getOwnDatastoreKind() {
+      return this.cls_.getOwnClassDatastoreKind();
+    },
+    function getDatastoreKind() {
+      return this.cls_.getClassDatastoreKind();
+    },
+    function getOwnDatastoreKey() {
+      if ( ! ( this.model_.ids || this.id ) ) {
+        throw new Error('Attempt to construct datastore key from ' +
+            'unidentified object');
+      }
+      var ids = this.model_.ids;
+      if ( ! ids )
+        return { kind: this.getOwnDatastoreKind(), name: this.id.toString() };
+
+      var name = new Array(ids.length);
+      for ( var i = 0; i < ids.length; i++ ) {
+        name = ids[i] + ( i < ids.length - 1 ? ':' : '' );
+      }
+      return { kind: this.getOwnDatastoreKind(), name: name };
+    },
+    function getDatastoreKey(opt_propertyPath) {
+      if ( ! opt_propertyPath ) return { path: [ this.getOwnDatastoreKey() ] };
+
+      var o = this;
+      var path = new Array(opt_propertyPath.length + 1);
+
+      path[0] = this.getOwnDatastoreKey();
+
+      for ( var i = 0; i < opt_propertyPath.length; i++ ) {
+        var next = opt_propertyPath[i];
+        var o = ( typeof next === 'string' ) ? o[next] : next.f(o);
+
+        if ( ! foam.core.FObject.isInstance(o) ) {
+          throw new Error('Attempt to get datastore key from non-keyable ' +
+              'object');
+        }
+
+        path[i + 1] = o.getOwnDatastoreKey();
+      }
+
+      return { path: path };
+    },
+    function toDatastoreEntity() {
+      var properties = {};
+      var ps = this.cls_.getAxiomsByClass(foam.core.Property);
+
+      for ( var i = 0; i < ps.length; i++ ) {
+        // TODO(markdittmer): MLangs that refer to storageTransient
+        // properties could cause the DatastoreDAO to misbehave. This could be
+        // fixed by auditing predicates and throwing an error when they contain
+        // properties that are dropped by the DAO processing the predicate.
+        if ( ps[i].storageTransient ) continue;
+
+        var value = ps[i].f(this);
+        properties[ps[i].name] = com.google.cloud.datastore.toDatastoreValue(
```

Comment:
Store com.google.cloud.datastore.toDatastoreValue in a local variable to avoid looking up on each iteration of the loop.